src/vs/vscode.d.ts

```diff
@@ -5221,6 +5246,17 @@ declare module 'vscode' {
 		 * @return Children of `element` or root if no element is passed.
 		 */
 		getChildren(element?: T): ProviderResult<T[]>;
+
+		/**
+		 * Optional method to return the parent of `element`.
+		 * Return `null` or `undefined` if `element` is a child of root.
+		 *
+		 * **NOTE:** This method should be implemented in order to access [reveal](#TreeView.reveal) API.
+		 *
+		 * @param element The element for which the parent has to be returned.
+		 * @return Parent of `element`.
+		 */
+		getParent?(element: T): ProviderResult<T>;
```

Comment:
Why don't we know this? We ask elements (parents) for their children, right? Couldn't we then keep/store that?

Reply:
Yes, I have the hierarchy. This is needed to support revealing an element that is not yet been asked. Since it is a pull model, we cannot expect extension to remember being loaded or not.

Reply:
Hm... I see.

Reply:
@sandy081 Sorry to jump in, I'm just playing with this API in insiders... How does it match up the nodes returns from getParent with those already in the tree? Eg. in your FTP sample getParent returns `{ resource: parent, isDirectory: true }` but I'm not sure how Code knows when it already has a node that matches that one to stop walking up?

Reply:
@DanTup 

> How does it match up the nodes returns from getParent with those already in the tree? Eg. in your FTP sample getParent returns { resource: parent, isDirectory: true } but I'm not sure how Code knows when it already has a node that matches that one to stop walking up?

I generated an `ID` for each element from its `TreeNode` representation. If `TreeNode` provides value for `id` property, then it is used. Otherwise, it is generated from its `label`. In the example you mentioned, I get the `ID` from the value `getParent` returns and get the hierarchy from the internal cache.

> I kept a full list of my tree nodes and returned the exact parent from here, but it seems to crash at the point I return undefined. If this is heavily in dev, I'll leave it a little while; but if you think it should work and want me to keep trying in insiders, let me know!

It should work in insiders and dev. When you return an `undefined` value for `getParent` it means the the requested element is one of the roots. So this should be handled without crash. Can you please file a separate issue for this and provide me a sample to reproduce?

Reply:
ID is unique per tree.

Reply:
Excellent, thanks!